src/cds/src/main/kotlin/org/icpclive/cds/cats/CATSDataSource.kt

```diff
@@ -112,18 +116,45 @@ internal class CATSDataSource(val settings: CatsSettings) : FullReloadContestDat
     private val problemsLoader = jsonLoader<Problems>(networkSettings = settings.network) { "${settings.url}/problems?cid=${settings.cid}&sid=${sid!!}&rows=1000&json=1" }
     private val usersLoader = jsonLoader<Users>(networkSettings = settings.network) { "${settings.url}/users?cid=${settings.cid}&sid=${sid!!}&rows=1000&json=1" }
     private val contestLoader = jsonLoader<Contest>(networkSettings = settings.network) { "${settings.url}/contest_params?cid=${settings.cid}&sid=${sid!!}&json=1" }
-    private val runsLoader = jsonLoader<List<Run>>(networkSettings = settings.network) { "${settings.url}/console?cid=${settings.cid}&sid=${sid!!}&rows=1000&json=1&search=is_ooc%3D0&show_messages=0&show_contests=0&show_results=1" }
+    private val runsLoader = jsonLoader<List<Run>>(networkSettings = settings.network) { "${settings.url}/console?cid=${settings.cid}&sid=${sid!!}&rows=1000&json=1&search=is_ooc%3D0&show_messages=0&show_contests=0&show_results=1&page=$page" }
 
     override suspend fun loadOnce(): ContestParseResult {
         sid = authLoader.load().sid
         return parseAndUpdateStandings(
             problemsLoader.load(),
             usersLoader.load(),
             contestLoader.load(),
-            runsLoader.load()
+            parseSubmitPages()
         )
     }
 
+    private suspend fun parseSubmitPages(): List<Run> {
+        val runs = mutableSetOf<Int>()
+        val result = mutableListOf<Run>()
+        for (currentPage in 0 until MAX_PAGE) {
+            page = currentPage
+            val pageSubmits = runsLoader.load().filterIsInstance<Submit>()
+
+            if (runs.intersect(pageSubmits.map(Submit::id).toSet()).size == pageSubmits.size) {
```

Comment:
I'm not sure about complexity of this operation, it looks quadratic overall for me (not sure, maybe there is a choice of smallest one of two inside). But anyway, maybe something like `pageSubmits.all { it.id in runs }` instead?

Reply:
Now it's working for O(n)